asmrun/amd64.S

```diff
@@ -122,95 +114,88 @@
 #endif
 
 /* Push the current exception handler.
-   (Stored as an offset, to survive stack reallocations). Clobbers REG */
-#define PUSH_EXN_HANDLER(REG) \
-        movq	Caml_state(exn_handler), REG; \
-        pushq   REG; CFI_ADJUST(8); \
+   (Stored as an offset, to survive stack reallocations). Clobbers %r11 */
+#define PUSH_EXN_HANDLER \
+        movq	Caml_state(exn_handler), %r11; \
+        pushq   %r11; CFI_ADJUST(8); \
         subq    %rsp, 0(%rsp)
 
-/* Pop the current exception handler. Undoes PUSH_EXN_HANDLER. Clobbers REG. */
-#define POP_EXN_HANDLER(REG) \
-        leaq	Caml_state(exn_handler), REG; \
+/* Pop the current exception handler. Undoes PUSH_EXN_HANDLER. Clobbers %r11. */
+#define POP_EXN_HANDLER \
+        leaq	Caml_state(exn_handler), %r11; \
         addq    %rsp, 0(%rsp); \
-        popq    (REG); CFI_ADJUST(-8)
+        popq    (%r11); CFI_ADJUST(-8)
 
 /******************************************************************************/
 /* Stack switching operations */
 /******************************************************************************/
 
-/* Switch from OCaml to C stack. Clobbers REG. */
-#define SWITCH_OCAML_TO_C_NO_CTXT(REG) \
-    /* Save OCaml SP in the stack slot */ \
-        movq	Caml_state(current_stack), REG; \
-        movq    %rsp, Stack_sp(REG); \
-    /* Switch to system stack */ \
-        movq	Caml_state(system_sp), REG; \
-        movq    %rsp, 16(REG); /* For DWARF CFI. See OFFXXX. */ \
-        movq    REG, %rsp; \
-        .cfi_def_cfa_offset 8 /* SYSCFAXXX */
+/* Switch from OCaml to C stack. Clobbers %r10, %r11.
+ * PUSHED is the number of bytes already pushed by this function.
+ * It is used to compute the backtrace. */
+#define SWITCH_OCAML_TO_C_NO_CTXT(PUSHED)                  \
+    /* Fill in Caml_state->current_stack->sp */            \
+        movq	Caml_state(current_stack), %r10;           \
+        movq    %rsp, Stack_sp(%r10);                      \
+    /* Fill in Caml_state->c_stack */                      \
+        movq	Caml_state(c_stack), %r11;                 \
+        movq    %rsp, Cstack_sp(%r11);                     \
+        movq    %r10, Cstack_stack(%r11);                  \
+    /* Switch to C stack */                                \
+        movq    %r11, %rsp;                                \
+        .cfi_remember_state;                               \
+        .cfi_escape DW_CFA_def_cfa_expression, 5,          \
```

Comment:
Assuming lib-unwind test will ensure that the DWARF is correct. :-)

Reply:
Yep, if this was wrong unwind_test wouldn't get past `caml_c_call`.